### DISCUSSION TITLE
Add support to export HAP (HarmonyOS Ability Package)

### DISCUSSION THREAD

**Discussion by lamdeyton** [godotengine/godot-proposals#10067] [Platforms]
### Describe the project you are working on

a 3d tool app in production

### Describe the problem or limitation you are having in your project

the tool app need to be able to run on Harmony Next OS.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add support to export HAP (HarmonyOS Ability Package)  which can be run on Harmony Next OS.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Export HAP  binary

### If this enhancement will not be used often, can it be worked around with a few lines of script?

no

### Is there a reason why this should be core and not an add-on in the asset library?

This will bring more users to godot

**Comment by Calinou** [comment]
Moving to discussion, as there is no concrete technical implementation provided.

Note that HarmonyOS has nothing in common in Android, so I expect a port to be nontrivial. Godot relies on a lot of different OS-specific APIs and requires hardware acceleration (OpenGL ES 3.0 or Vulkan 1.0[^1]).

[^1]: A lot of Android devices have buggy drivers (particularly those limited to Vulkan 1.0 or 1.1), and I expect this to transfer to HarmonyOS. In practice, it's strongly recommended to have Vulkan 1.2 support.

**Reply by xmxy-wt** [reply]
Here's the translation of your suggestion:

I hope that support for exporting "HarmonyOS" programs can be added.

I believe the Chinese market is still vast. However, for the Chinese market, Godot cannot export to WeChat Mini Programs and Douyin Mini Programs, which is quite disappointing. If HarmonyOS continues to be unsupported, a large number of developers will turn to other engines.

**Reply by octanejohn** [reply]
"mini programs" is a custom/proprietary web-like standard using json , custom css(wxss) and javascript so better to focus on web apps

harmony os next is like android with java/c++ but ts/cangjie with own their runtime

all apis are private/proprietary, have to accept their licenses to even access it so is pointless unless they become as worldwide as apple

**Reply by dsyze** [reply]
> Moving to discussion, as there is no concrete technical implementation provided.转向讨论，因为没有提供具体的技术实现。
> 
> Note that HarmonyOS has nothing in common in Android, so I expect a port to be nontrivial. Godot relies on a lot of different OS-specific APIs and requires hardware acceleration (OpenGL ES 3.0 or Vulkan 1.0[1](#user-content-fn-1-d342424f675c27a06b8694d79929701f)).请注意，HarmonyOS 与 Android 没有任何共同点，因此我预计移植会很重要。 Godot 依赖于许多不同的特定于操作系统的 API，并且需要硬件加速（OpenGL ES 3.0 或 Vulkan 1.0 [1](#user-content-fn-1-d342424f675c27a06b8694d79929701f) ）。
> 
> ## Footnotes
> 1. A lot of Android devices have buggy drivers (particularly those limited to Vulkan 1.0 or 1.1), and I expect this to transfer to HarmonyOS. In practice, it's strongly recommended to have Vulkan 1.2 support. [↩](#user-content-fnref-1-d342424f675c27a06b8694d79929701f)许多 Android 设备都有错误的驱动程序（特别是仅限于 Vulkan 1.0 或 1.1 的驱动程序），我预计这会转移到 HarmonyOS。在实践中，强烈建议支持 Vulkan 1.2。 [↩](#user-content-fnref-1-d342424f675c27a06b8694d79929701f)

Will this work be included in Godot’s official roadmap in the short term? If not, could you provide some key suggestions for the porting process to help the community complete this work more smoothly?

**Reply by AThousandShips** [reply]
What work are you talking about?

**Reply by dsyze** [reply]
> What work are you talking about?你说的是什么工作？

Add support to export HAP (HarmonyOS Ability Package)

**Reply by AThousandShips** [reply]
There's no details on how to do this provided so we can't plan to do something we don't know is possible, or even wanted

So more information is needed. Like if this is possible with open source

**Reply by dsyze** [reply]
> There's no details on how to do this provided so we can't plan to do something we don't know is possible, or even wanted
> 
> So more information is needed. Like if this is possible with open source

Thanks, as for now, the Chinese agent of the Unity engine has completed export support for HarmonyOS. I hope the information in the following link will be helpful: https://docs.unity.cn/cn/tuanjiemanual/Manual/openharmony-introducing.html

**Reply by dsyze** [reply]
> There's no details on how to do this provided so we can't plan to do something we don't know is possible, or even wanted
> 
> So more information is needed. Like if this is possible with open source

I’d like to start working on the HarmonyOS port. Is there an existing workflow I can reference? The development docs for porting to Android should be quite helpful.

**Reply by Calinou** [reply]
> Is there an existing workflow I can reference?

The [Custom platform ports](https://docs.godotengine.org/en/stable/contributing/development/core_and_modules/custom_platform_ports.html) documentation is a good starting point.

How does OpenGL ES and Vulkan support look like on HarmonyOS? The Forward+/Mobile rendering methods require Vulkan support, while Compatibility requires OpenGL ES 3.0.

**Reply by waci11** [reply]
> > Is there an existing workflow I can reference?
> 
> The [Custom platform ports](https://docs.godotengine.org/en/stable/contributing/development/core_and_modules/custom_platform_ports.html) documentation is a good starting point.
> 
> How does OpenGL ES and Vulkan support look like on HarmonyOS? The Forward+/Mobile rendering methods require Vulkan support, while Compatibility requires OpenGL ES 3.0.

[https://developer.huawei.com/consumer/en/doc/harmonyos-guides-V5/arkgraphics3d-overview-V5#constraints](url)
To use the ArkGraphics 3D service, the device must support the GPU driver of OpenGL ES 3.2 or later or Vulkan 1.0 or later.

I think it support OpenGL ES 3.2 or later or Vulkan 1.0 .

**Reply by dsyze** [reply]
> > Is there an existing workflow I can reference?
> 
> The [Custom platform ports](https://docs.godotengine.org/en/stable/contributing/development/core_and_modules/custom_platform_ports.html) documentation is a good starting point.
> 
> How does OpenGL ES and Vulkan support look like on HarmonyOS? The Forward+/Mobile rendering methods require Vulkan support, while Compatibility requires OpenGL ES 3.0.

OK, I have a guy who is skilled at C++, and he told he wants to make games by himself in the future. I introduced him Godot, and now he has started learning Godot. Trying to make godot Porting to HarmonyOs must be a good practice for him.

**Reply by kdada** [reply]
Any progress?

I might try to port Godot to OpenHarmony (or its distribution HarmonyOS Next) at some point in the future, but it's not easy and I'd like to see if anyone in the community is already working on it.

**Reply by Calinou** [reply]
Proposal opened:

- https://github.com/godotengine/godot-proposals/issues/12734

**Comment by zhudisha** [comment]
I really hope that Godot can adapt to OpenHarmony in the future

**Comment by LuxusIX** [comment]
This is a crucial initiative. Since the complete separation between Google (Android, Play Store) and Huawei, the Chinese tech giant has had no choice but to build its own ecosystem from the ground up, and its success has been nothing short of staggering.
As of 2026, **HarmonyOS now powers over 1 billion devices**. With an massive user base and the **AppGallery** store rapidly expanding, this platform represents a significant, untapped opportunity for the Godot community. Adding support for this OS is the key to ensuring Godot remains a premier choice for developers looking to publish their games on this massive, growing market.
#### The Technical Challenge
I am fully aware that the technical hurdles are immense. We are talking about re-engineering core components of the engine to interface with a system that deviates fundamentally from existing standards (specifically regarding **ArkUI**, the native application lifecycle management, and N-API requirements).
#### Why it's worth the effort
Despite the complexity, I believe the effort is justified for several reasons:
 * **Strategic Growth:** It allows Godot to capture a major share of an ecosystem that is currently underserved by open-source engines.
 * **Core Philosophy:** True to the "Godot Everywhere" philosophy, ignoring a third of the global mobile device market contradicts our goal of being a truly universal engine.
 * **Community Potential:** Enabling this export target will empower developers to bring their creative projects to millions of new players.
I believe that even a phased, MVP-style implementation would be a game-changer. I am eager to see how we can support the contributors working on this to make it a reality.
